janusgraph-cql/src/main/java/org/janusgraph/diskstorage/cql/builder/CQLSessionBuilder.java

```diff
@@ -49,50 +54,91 @@ public CqlSession build(Configuration configuration, String[] baseHostnames, int
             contactPoints.add(contactPoint);
         }
 
+        DriverConfigLoader driverConfigLoader;
+        if(configuration.get(CQLConfigOptions.BASE_PROGRAMMATIC_CONFIGURATION_ENABLED)){
+            driverConfigLoader = baseConfigurationLoaderBuilder.build(configuration, contactPoints, baseConnectionTimeoutMS);
+        } else {
+            driverConfigLoader = null;
+        }
 
+        Optional<Supplier<Config>> internalConfigurationSupplier = getInternalConfigSupplier(configuration, driverConfigLoader == null);
         final CqlSessionBuilder builder = CqlSession.builder();
 
-        Stack<DriverConfigLoader> driverConfigLoadersToUse = new Stack<>();
+        if(internalConfigurationSupplier.isPresent()){
+            DriverConfigLoader internalDriverConfigLoader = new DefaultDriverConfigLoader(
+                internalConfigurationSupplier.get(), false);
+            if(driverConfigLoader == null){
+                driverConfigLoader = internalDriverConfigLoader;
+            } else {
+                driverConfigLoader = DriverConfigLoader.compose(internalDriverConfigLoader, driverConfigLoader);
+            }
+        }
 
-        if(configuration.get(CQLConfigOptions.BASE_PROGRAMMATIC_CONFIGURATION_ENABLED)){
-            driverConfigLoadersToUse.push(baseConfigurationLoaderBuilder.build(configuration, contactPoints, baseConnectionTimeoutMS));
+        if(driverConfigLoader != null){
+            builder.withConfigLoader(driverConfigLoader);
+        }
+
+        return builder.build();
+    }
+
+    private Optional<Supplier<Config>> getInternalConfigSupplier(Configuration configuration, boolean withFallbackDefaultConfiguration) throws PermanentBackendException {
+        boolean hasFileConfiguration = configuration.has(CQLConfigOptions.FILE_CONFIGURATION);
+        boolean hasResourceConfiguration = configuration.has(CQLConfigOptions.RESOURCE_CONFIGURATION);
+        boolean hasStringConfiguration = configuration.has(CQLConfigOptions.STRING_CONFIGURATION);
+        boolean hasUrlConfiguration = configuration.has(CQLConfigOptions.URL_CONFIGURATION);
+
+        boolean hasAnyInternalConfiguration = hasFileConfiguration
+            || hasResourceConfiguration
+            || hasStringConfiguration
+            || hasUrlConfiguration;
+
+        if(!hasAnyInternalConfiguration){
+            return Optional.empty();
```

Comment:
Just trying to understand: Why do we return empty even if `withFallbackDefaultConfiguration` is true?

Reply:
A good question. I decided to preserve our current builder logic. If no configuration is used (i.e. `hasAnyInternalConfiguration == true && BASE_PROGRAMMATIC_CONFIGURATION_ENABLED == false)`) then we don't provide any configuration into the builder which leads to use default CQL configuration options

Reply:
Here it checks, if we didn't provide any configuration our-self then it uses default configuration options: https://github.com/datastax/java-driver/blob/6304cb725a858e107ac826edfa6c8a6983504d78/core/src/main/java/com/datastax/oss/driver/api/core/session/SessionBuilder.java#L849